README.md
# mp-yearmonth

A year-month datatype for Python.

## Installation

```bash
pip install mp-yearmonth
```

## Usage

```python
from mp_yearmonth import YearMonth

ym = YearMonth(2019, 1)
print(ym) # 2019-01
```

Parsing ISO 8601 strings:

```python
ym = YearMonth.parse("2019-01")

print(ym.year) # 2019
print(ym.month) # 1
```

Comparisons:

```python
ym1 = YearMonth(2019, 1)
ym2 = YearMonth(2019, 2)

print(ym1 < ym2) # True
```

Addition:

```python
ym = YearMonth(2019, 1)
ym += 1

print(ym) # 2019-02
```

Range:

```python
ym1 = YearMonth(2019, 1)
ym2 = YearMonth(2019, 3)

for ym in YearMonth.range(ym1, ym2):
    print(ym) # 2019-01, 2019-02, 2019-03
```

## License

mp-yearmonth is licensed under the MIT license. See [LICENSE](https://github.com/raymondjavaxx/mp-yearmonth/blob/main/LICENSE) for details.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

## v0.2.0

### Added

- Added support for membership test operators (`in` and `not in`) for checking if a `date` or `datetime` falls within a `YearMonth`.

### Changed

- Improved `YearMonth.parse()` performance.

## v0.1.0

- Initial release.


tests/test_yearmonth.py
import pytest
import datetime
from mp_yearmonth import YearMonth


def test_init_handles_out_of_range_month():
    with pytest.raises(ValueError, match=r"^month must be between 1 and 12$"):
        YearMonth(2021, 0)

    with pytest.raises(ValueError, match=r"^month must be between 1 and 12$"):
        YearMonth(2021, 13)


def test_init_handles_out_of_range_year():
    with pytest.raises(ValueError, match=r"^year must be between 1 and 9999$"):
        YearMonth(0, 1)

    with pytest.raises(ValueError, match=r"^year must be between 1 and 9999$"):
        YearMonth(10000, 1)


def test_init_handles_non_int_values():
    with pytest.raises(TypeError, match=r"^year must be an integer$"):
        YearMonth("2021", 1)

    with pytest.raises(TypeError, match=r"^month must be an integer$"):
        YearMonth(2021, "1")


def test_str():
    assert str(YearMonth(2021, 1)) == "2021-01"
    assert str(YearMonth(2021, 12)) == "2021-12"
    assert str(YearMonth(100, 1)) == "0100-01"


def test_iso8601():
    assert YearMonth(2021, 1).iso8601 == "2021-01"
    assert YearMonth(2021, 12).iso8601 == "2021-12"
    assert YearMonth(100, 1).iso8601 == "0100-01"


def test_repr():
    assert repr(YearMonth(2021, 1)) == "YearMonth(2021, 1)"
    assert repr(YearMonth(2021, 12)) == "YearMonth(2021, 12)"


def test_eq():
    assert YearMonth(2021, 1) == YearMonth(2021, 1)
    assert YearMonth(2021, 1) != YearMonth(2021, 2)
    assert YearMonth(2021, 1) != object()


def test_lt():
    assert YearMonth(2021, 1) < YearMonth(2021, 2)
    assert YearMonth(2021, 1) < YearMonth(2022, 1)
    assert not (YearMonth(2021, 1) < YearMonth(2021, 1))
    assert not (YearMonth(2021, 1) < YearMonth(2020, 12))


def test_le():
    assert YearMonth(2021, 1) <= YearMonth(2021, 2)
    assert YearMonth(2021, 1) <= YearMonth(2022, 1)
    assert YearMonth(2021, 1) <= YearMonth(2021, 1)
    assert not (YearMonth(2021, 1) <= YearMonth(2020, 12))


def test_gt():
    assert YearMonth(2021, 2) > YearMonth(2021, 1)
    assert YearMonth(2022, 1) > YearMonth(2021, 1)
    assert not (YearMonth(2021, 1) > YearMonth(2021, 1))
    assert not (YearMonth(2020, 12) > YearMonth(2021, 1))


def test_ge():
    assert YearMonth(2021, 2) >= YearMonth(2021, 1)
    assert YearMonth(2022, 1) >= YearMonth(2021, 1)
    assert YearMonth(2021, 1) >= YearMonth(2021, 1)
    assert not (YearMonth(2020, 12) >= YearMonth(2021, 1))


def test_hash():
    assert hash(YearMonth(2021, 1)) == hash(YearMonth(2021, 1))  # Same
    assert hash(YearMonth(2021, 1)) != hash(YearMonth(2021, 2))  # Different month
    assert hash(YearMonth(2022, 1)) != hash(YearMonth(2021, 1))  # Different year


def test_numdays():
    assert YearMonth(2021, 1).numdays == 31
    assert YearMonth(2021, 2).numdays == 28
    assert YearMonth(2021, 3).numdays == 31
    assert YearMonth(2021, 4).numdays == 30
    assert YearMonth(2021, 5).numdays == 31
    assert YearMonth(2021, 6).numdays == 30
    assert YearMonth(2021, 7).numdays == 31
    assert YearMonth(2021, 8).numdays == 31
    assert YearMonth(2021, 9).numdays == 30
    assert YearMonth(2021, 10).numdays == 31
    assert YearMonth(2021, 11).numdays == 30
    assert YearMonth(2021, 12).numdays == 31

    # Leap year
    assert YearMonth(2020, 2).numdays == 29


def test_current():
    assert YearMonth.current() == YearMonth.parse(str(YearMonth.current()))


def test_parse():
    assert YearMonth.parse("2021-01") == YearMonth(2021, 1)
    assert YearMonth.parse("2021-12") == YearMonth(2021, 12)


def test_parse_handles_invalid_input():
    with pytest.raises(ValueError, match=r"^Invalid YearMonth string format"):
        YearMonth.parse("2021-1")

    with pytest.raises(ValueError, match=r"^Invalid YearMonth string format"):
        YearMonth.parse("21-01")
        assert "Invalid YearMonth string format"

    with pytest.raises(ValueError, match=r"^Invalid YearMonth string format"):
        YearMonth.parse("2021-1a")
        assert "Invalid YearMonth string format"

    with pytest.raises(ValueError, match=r"^Invalid YearMonth string format"):
        YearMonth.parse("2021-01-01")
        assert "Invalid YearMonth string format"

    with pytest.raises(ValueError, match=r"^month must be between 1 and 12$"):
        YearMonth.parse("2021-13")
        assert "Invalid YearMonth string format"


def test_next():
    assert YearMonth(2021, 1).next() == YearMonth(2021, 2)
    assert YearMonth(2021, 12).next() == YearMonth(2022, 1)


def test_range():
    # Ascending
    r = YearMonth.range(YearMonth(2021, 1), YearMonth(2021, 3))
    assert list(r) == [YearMonth(2021, 1), YearMonth(2021, 2), YearMonth(2021, 3)]

    # Descending
    r = YearMonth.range(YearMonth(2021, 3), YearMonth(2021, 1))
    assert list(r) == [YearMonth(2021, 3), YearMonth(2021, 2), YearMonth(2021, 1)]

    # Same
    r = YearMonth.range(YearMonth(2021, 1), YearMonth(2021, 1))
    assert list(r) == [YearMonth(2021, 1)]


def test_applying_delta():
    ym = YearMonth(2021, 1)
    assert ym.applying_delta(0) == YearMonth(2021, 1)
    assert ym.applying_delta(1) == YearMonth(2021, 2)
    assert ym.applying_delta(2) == YearMonth(2021, 3)
    assert ym.applying_delta(3) == YearMonth(2021, 4)
    assert ym.applying_delta(4) == YearMonth(2021, 5)
    assert ym.applying_delta(5) == YearMonth(2021, 6)
    assert ym.applying_delta(6) == YearMonth(2021, 7)
    assert ym.applying_delta(7) == YearMonth(2021, 8)
    assert ym.applying_delta(8) == YearMonth(2021, 9)
    assert ym.applying_delta(9) == YearMonth(2021, 10)
    assert ym.applying_delta(10) == YearMonth(2021, 11)
    assert ym.applying_delta(11) == YearMonth(2021, 12)
    assert ym.applying_delta(12) == YearMonth(2022, 1)

    assert ym.applying_delta(-1) == YearMonth(2020, 12)
    assert ym.applying_delta(-2) == YearMonth(2020, 11)


def test_add():
    ym = YearMonth(2021, 1)
    assert ym + 0 == YearMonth(2021, 1)
    assert ym + 1 == YearMonth(2021, 2)
    assert ym + 2 == YearMonth(2021, 3)


def test_sub():
    ym = YearMonth(2021, 1)
    assert ym - 0 == YearMonth(2021, 1)
    assert ym - 1 == YearMonth(2020, 12)
    assert ym - 2 == YearMonth(2020, 11)


def test_contains_handles_dates():
    assert datetime.date(2021, 1, 1) in YearMonth(2021, 1)
    assert datetime.date(2021, 1, 31) in YearMonth(2021, 1)
    assert datetime.date(2021, 2, 1) not in YearMonth(2021, 1)


def test_contains_handles_datetimes():
    assert datetime.datetime(2021, 1, 1, hour=12, minute=30) in YearMonth(2021, 1)
    assert datetime.datetime(2021, 1, 31, hour=12, minute=30) in YearMonth(2021, 1)
    assert datetime.datetime(2021, 2, 1, hour=12, minute=30) not in YearMonth(2021, 1)


def test_contains_handles_unsupported_types():
    assert 1 not in YearMonth(2021, 1)
    assert 2021 not in YearMonth(2021, 1)
    assert "test" not in YearMonth(2021, 1)


def test_bounds():
    ym1 = YearMonth(2021, 1)
    assert ym1.bounds() == (datetime.date(2021, 1, 1), datetime.date(2021, 1, 31))

    ym2 = YearMonth(2021, 2)
    assert ym2.bounds() == (datetime.date(2021, 2, 1), datetime.date(2021, 2, 28))


mp_yearmonth/yearmonth.py
# yearmonth.py
# Copyright (c) 2023 Ramon Torres
#
# This module is part of mp-yearmonth and is released under
# the MIT License: https://opensource.org/license/mit/

import re
import calendar
import datetime
from typing import Iterator, Tuple

_FORMAT_PATTERN = re.compile(r"^(\d{4})-(\d{2})$")


class YearMonth:
    """A class representing a year and month.

    It can represent any month between 0001-01 and 9999-12:

            >>> ym = YearMonth(2021, 1)

    You can use this class to iterate over a range of months:

            >>> start = YearMonth(2021, 1)
            >>> end = YearMonth(2021, 3)
            >>> for ym in YearMonth.range(start, end):
            ...     print(ym)
            2021-01
            2021-02
            2021-03

    You can also add and subtract months:

            >>> ym = YearMonth(2021, 1)
            >>> next_month = ym + 1
            >>> print(next_month)
            2021-02

    Attributes:
        year (int)  : The year.
        month (int) : The month.
    """

    def __init__(self, year: int, month: int):
        """Create a new YearMonth object.

        Args:
            year: The year.
            month: The month."""
        if not isinstance(year, int):
            raise TypeError("year must be an integer")

        if not isinstance(month, int):
            raise TypeError("month must be an integer")

        if not 1 <= month <= 12:
            raise ValueError("month must be between 1 and 12")

        if not datetime.MINYEAR <= year <= datetime.MAXYEAR:
            raise ValueError(
                f"year must be between {datetime.MINYEAR} and {datetime.MAXYEAR}"
            )

        self.year = year
        self.month = month

    def __str__(self):
        return self.iso8601

    def __repr__(self):
        return f"YearMonth({self.year}, {self.month})"

    def __eq__(self, other):
        if isinstance(other, YearMonth):
            return self.year == other.year and self.month == other.month
        return NotImplemented

    def __lt__(self, other):
        if isinstance(other, YearMonth):
            return (self.year, self.month) < (other.year, other.month)
        return NotImplemented

    def __le__(self, other):
        if isinstance(other, YearMonth):
            return (self.year, self.month) <= (other.year, other.month)
        return NotImplemented

    def __gt__(self, other):
        if isinstance(other, YearMonth):
            return (self.year, self.month) > (other.year, other.month)
        return NotImplemented

    def __ge__(self, other):
        if isinstance(other, YearMonth):
            return (self.year, self.month) >= (other.year, other.month)
        return NotImplemented

    def __hash__(self):
        return hash((self.year, self.month))

    def __add__(self, other) -> "YearMonth":
        if isinstance(other, int):
            return self.applying_delta(other)
        return NotImplemented

    def __sub__(self, other) -> "YearMonth":
        if isinstance(other, int):
            return self.applying_delta(-other)
        return NotImplemented

    def __contains__(self, other) -> bool:
        if isinstance(other, datetime.date) or isinstance(other, datetime.datetime):
            return self.year == other.year and self.month == other.month
        return False

    # Public interface

    @property
    def iso8601(self) -> str:
        """The ISO 8601 representation of the year and month."""
        return f"{self.year:04d}-{self.month:02d}"

    @property
    def numdays(self) -> int:
        """The number of days in the month."""
        _, days = calendar.monthrange(self.year, self.month)
        return days

    @classmethod
    def current(cls) -> "YearMonth":
        """Return the current year and month."""
        now = datetime.date.today()
        return cls(now.year, now.month)

    @classmethod
    def parse(cls, s: str) -> "YearMonth":
        """Parses a string in the format YYYY-MM into a YearMonth object."""
        match = _FORMAT_PATTERN.match(s)
        if not match:
            raise ValueError(f"Invalid YearMonth string format: {s}")

        year, month = match.groups()
        return cls(int(year), int(month))

    def next(self) -> "YearMonth":
        """Return the next month."""
        if self.month == 12:
            return YearMonth(self.year + 1, 1)
        return YearMonth(self.year, self.month + 1)

    def prev(self) -> "YearMonth":
        """Return the previous month."""
        if self.month == 1:
            return YearMonth(self.year - 1, 12)
        return YearMonth(self.year, self.month - 1)

    def bounds(self) -> Tuple[datetime.date, datetime.date]:
        """Return the first and last day of the month."""
        return (
            datetime.date(self.year, self.month, 1),
            datetime.date(self.year, self.month, self.numdays),
        )

    def applying_delta(self, months: int) -> "YearMonth":
        """Return the YearMonth that is `months` away from this one."""
        year = self.year + ((self.month + months - 1) // 12)
        month = ((self.month + months - 1) % 12) + 1
        return YearMonth(year, month)

    @classmethod
    def range(cls, start: "YearMonth", end: "YearMonth") -> Iterator["YearMonth"]:
        """Return a list of all months between start and end, inclusive."""
        is_ascending = start <= end
        current = start
        while current != end:
            yield current
            current = current.next() if is_ascending else current.prev()
        yield end
